### PR TITLE
Refact dec

### PR DESCRIPTION
**Pull request by mengker33**

### CODE REVIEW

**Comment by mengker33** [comment]
@xhaihao, @HeJunyan, @tong1wu, @liuyinhangx, @xuguangxin, please help to review. Thank you :)

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 1]
```diff
@@ -286,6 +289,73 @@ get_surface (GstMsdkDec * thiz, GstBuffer * buffer)
...
+  gst_msdkdec_free_unlocked_msdk_surfaces (thiz);
+#ifndef _WIN32
+  if ((gst_buffer_pool_acquire_buffer (thiz->alloc_pool, &out_buffer, 0))
+      != GST_FLOW_OK) {
+    GST_ERROR_OBJECT (thiz, "Failed to allocate output buffer");
+    return NULL;
+  }
+
+  GstMemory *mem = NULL;
+  mfxFrameSurface1 *mfx_surface = NULL;
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:238]: Mixed declaration?

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 2]
```diff
@@ -286,6 +289,73 @@ get_surface (GstMsdkDec * thiz, GstBuffer * buffer)
...
+  msdk_surface =
+      gst_msdk_import_sys_mem_to_msdk_surface (out_buffer, output_state->info);
+#endif
+
+  if (msdk_surface) {
+    msdk_surface->buf = out_buffer;
+  }
+
+  if (thiz->use_video_memory && !msdk_surface->from_qdata) {
+    GST_ERROR ("No available surfaces");
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:274]: You should free msdk_surface here

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 3]
```diff
@@ -286,6 +289,73 @@ get_surface (GstMsdkDec * thiz, GstBuffer * buffer)
...
+
+  mem = gst_buffer_peek_memory (out_buffer, 0);
+  msdk_surface = g_slice_new0 (GstMsdkSurface);
+  msdk_surface->from_qdata = FALSE;
+
+  if ((mfx_surface = gst_mini_object_get_qdata (GST_MINI_OBJECT_CAST (mem),
+              GST_MSDK_FRAME_SURFACE))) {
+    msdk_surface->surface = mfx_surface;
+    msdk_surface->from_qdata = TRUE;
+  }
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:248]: Adding an else to handle mfx_surface == NULL and remove line 344 might be better

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c [hunk 4]
```diff
@@ -254,18 +153,21 @@ gst_msdk_frame_alloc (mfxHDL pthis, mfxFrameAllocRequest * req,
 
   gst_msdk_context_add_alloc_response (context, msdk_resp);
 
+  /* We need to put all the buffers back to the pool */
+  GList *l;
+  GstMsdkSurface *tmp = NULL;
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c:158]: Mixed declaration

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 5]
```diff
@@ -161,24 +161,72 @@ gst_msdkdec_free_unlocked_msdk_surfaces (GstMsdkDec * thiz)
...
+      thiz->param.mfx.FrameInfo.FrameRateExtN);
 
   /* Free un-unsed msdk surfaces firstly, hence the associated mfx
    * surfaces will be moved from used list to available list */
-  gst_msdkdec_free_unlocked_msdk_surfaces (thiz);
+  if (!gst_msdkdec_free_unlocked_msdk_surfaces (thiz)) {
+    while (n < retry_times) {
+      /* It is MediaSDK requirement that only the pre-allocated
+       * surfaces can be used during the whole decoding process.
+       * In the case of decoder plus multi-encdoers, it is possible
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:219]: multi-encoders

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c (hunk 2 above)
Comment (mengker33) [reply subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c]: > You should free msdk_surface here

Fixed.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c (hunk 3 above)
Comment (mengker33) [reply subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:253]: > Adding an else to handle mfx_surface == NULL and remove line 344 might be better

Thanks for the advice. Done.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c [hunk 6]
```diff
@@ -104,158 +91,61 @@ gst_msdk_frame_alloc (mfxHDL pthis, mfxFrameAllocRequest * req,
...
+      GstBuffer *buf;
 
-      status = gst_msdk_get_mfx_status_from_va_status (va_status);
-      if (status < MFX_ERR_NONE) {
-        GST_ERROR ("failed to create buffer");
-        return status;
+      if (gst_buffer_pool_acquire_buffer (pool, &buf, NULL) != GST_FLOW_OK) {
+        GST_ERROR ("Failed to allocate texture buffer");
+        gst_buffer_pool_set_active (pool, FALSE);
+        gst_object_unref (pool);
```
[the remark this answers is not in the document]
Comment (mengker33) [reply subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c:139]: > In this for loop, the pool may be deactived and unreferred multiple times.

Thanks for pointing this out. Fixed to return ERR status when "failed to allocate buffer".

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkcontext.c [hunk 7]
```diff
@@ -590,8 +591,6 @@ gst_msdk_context_add_alloc_response (GstMsdkContext * context,
 {
   context->priv->cached_alloc_responses =
       g_list_prepend (context->priv->cached_alloc_responses, resp);
-
-  create_surfaces (context, resp);
```
Comment (HeJunyan) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkcontext.c:594]: After delete this,  the create_surfaces() function seems no usage, please delete. Or the compiler will generate warning.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 8]
```diff
@@ -161,13 +161,43 @@ gst_msdkdec_free_unlocked_msdk_surfaces (GstMsdkDec * thiz)
...
+{
+  GList *l;
+  GstMsdkSurface *surface = NULL;
+  gboolean ret = FALSE;
+
+  for (l = thiz->locked_msdk_surfaces; l;) {
+    GList *next = l->next;
+    surface = l->data;
+    if (surface->surface->Data.Locked == 0
+        && GST_MINI_OBJECT_REFCOUNT (surface->buf) == 1) {
```
Comment (HeJunyan) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:191]: It's so tricky here. Please at lease use `GST_MINI_OBJECT_REFCOUNT_VALUE`, which is a atomic operation.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 9]
```diff
@@ -161,13 +161,43 @@ gst_msdkdec_free_unlocked_msdk_surfaces (GstMsdkDec * thiz)
...
+  /* We need to check if all surfaces are in used */
+  if (g_list_length (thiz->locked_msdk_surfaces) ==
+      thiz->alloc_resp.NumFrameActual)
+    return FALSE;
+  else
+    return TRUE;
+}
+
+static gboolean
+check_surface_available (GstMsdkDec * thiz)
```
Comment (HeJunyan) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:181]: It is very similar with the previous gst_msdkdec_free_unlocked_msdk_surfaces(), can you merge them together?

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 10]
```diff
@@ -179,10 +209,28 @@ allocate_output_surface (GstMsdkDec * thiz)
...
 #endif
+  gint n = 0;
+  guint retry_times = gst_util_uint64_scale_ceil (GST_USECOND,
+      thiz->param.mfx.FrameInfo.FrameRateExtD,
+      thiz->param.mfx.FrameInfo.FrameRateExtN);
 
   /* Free un-unsed msdk surfaces firstly, hence the associated mfx
    * surfaces will be moved from used list to available list */
-  gst_msdkdec_free_unlocked_msdk_surfaces (thiz);
+  if (!gst_msdkdec_free_unlocked_msdk_surfaces (thiz)) {
```
Comment (HeJunyan) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:219]: How about
while ((!gst_msdkdec_free_unlocked_msdk_surfaces (thiz) && n < retry_times) {
      n++;
      g_usleep (1000);
}

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkh265dec.c [hunk 11]
```diff
@@ -70,8 +77,8 @@ static GstStaticPadTemplate sink_factory = GST_STATIC_PAD_TEMPLATE ("sink",
 static GstStaticPadTemplate src_factory = GST_STATIC_PAD_TEMPLATE ("src",
     GST_PAD_SRC,
     GST_PAD_ALWAYS,
-    GST_STATIC_CAPS (GST_MSDK_CAPS_STR (COMMON_FORMAT, COMMON_FORMAT))
-    );
+    GST_STATIC_CAPS (GST_MSDK_CAPS_STR (COMMON_FORMAT, COMMON_FORMAT)
+        VA_SRC_CAPS_STR));
```
Comment (HeJunyan) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkh265dec.c:81]: How about  VP9 and Jpeg? I do not notice that you add this for them.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c (hunk 8 above)
Comment (mengker33) [reply subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:191]: > It's so tricky here. Please at lease use `GST_MINI_OBJECT_REFCOUNT_VALUE`, which is a atomic operation.

I can not think of another way to check if downstream has released this buffer before we let msdkdec unref it. Do you have a better solution?

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c (hunk 9 above)
Comment (mengker33) [reply subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c]: The func gst_msdkdec_free_unlocked_msdk_surfaces() aims to free all un-used msdk surfaces back to the pool before we acquire a buffer from the pool while check_surface_available() just needs to find one available surface. So, here I  merge this two and add a flag (check_avail_surface) to decide if we need to find all (flag is TRUE) or just one surface (flag is FALSE).

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkh265dec.c (hunk 11 above)
Comment (mengker33) [reply subprojects/gst-plugins-bad/sys/msdk/gstmsdkh265dec.c:81]: Added va caps for vp9 and jpeg.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c [hunk 12]
```diff
@@ -104,158 +91,62 @@ gst_msdk_frame_alloc (mfxHDL pthis, mfxFrameAllocRequest * req,
...
-
-      status = gst_msdk_get_mfx_status_from_va_status (va_status);
-      if (status < MFX_ERR_NONE) {
-        GST_ERROR ("failed to create buffer");
-        return status;
+      if (gst_buffer_pool_acquire_buffer (pool, &buf, NULL) != GST_FLOW_OK) {
+        GST_ERROR ("Failed to allocate buffer");
+        gst_buffer_pool_set_active (pool, FALSE);
+        gst_object_unref (pool);
+        return MFX_ERR_MEMORY_ALLOC;
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c:140]: Do we need to free the malloced memory before return?

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c [hunk 13]
```diff
@@ -104,158 +91,62 @@ gst_msdk_frame_alloc (mfxHDL pthis, mfxFrameAllocRequest * req,
...
+    gst_video_alignment_reset (&align);
+    gst_msdk_set_video_alignment
+        (&info, req->Info.Width, req->Info.Height, &align);
+    gst_video_info_align (&info, &align);
+
+    caps = gst_video_info_to_caps (&info);
+
+    pool = gst_msdk_context_get_alloc_pool (context);
+    if (!pool) {
+      return MFX_ERR_MEMORY_ALLOC;
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c:118]: Free the malloced memory?

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 14]
```diff
@@ -146,140 +146,132 @@ gst_msdkdec_get_oldest_frame (GstVideoDecoder * decoder)
...
+       * least one surface is free for decoder.
+       */
+      n++;
+      g_usleep (1000);
+      if (gst_msdkdec_free_unlocked_msdk_surfaces (thiz, TRUE))
+        break;
+    }
+  }
+#ifndef _WIN32
+  if ((gst_buffer_pool_acquire_buffer (thiz->alloc_pool, &out_buffer, 0))
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:220]: Better replace 0 with NULL.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 15]
```diff
@@ -146,140 +146,132 @@ gst_msdkdec_get_oldest_frame (GstVideoDecoder * decoder)
...
+  if (!gst_buffer_pool_is_active (thiz->pool) &&
+      !gst_buffer_pool_set_active (thiz->pool, TRUE)) {
+    GST_ERROR_OBJECT (thiz, "Failed to activate buffer pool");
+    return NULL;
   }
 
-  *buffer = gst_buffer_ref (frame->output_buffer);
-  gst_buffer_replace (&frame->output_buffer, NULL);
-  gst_video_codec_frame_unref (frame);
+  if ((gst_buffer_pool_acquire_buffer (thiz->pool, &out_buffer, 0))
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:246]: Same here.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 16]
```diff
@@ -146,140 +146,132 @@ gst_msdkdec_get_oldest_frame (GstVideoDecoder * decoder)
...
+      gst_video_decoder_get_output_state (GST_VIDEO_DECODER (thiz));
 
-static MsdkSurface *
-get_surface (GstMsdkDec * thiz, GstBuffer * buffer)
-{
-  MsdkSurface *i;
-  GstVideoCodecState *output_state = NULL;
-  gboolean success;
+  msdk_surface =
+      gst_msdk_import_sys_mem_to_msdk_surface (out_buffer, output_state->info);
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:256]: This is not right. The gst_msdk_import_sys_mem_to_msdk_surface returns a malloced msdk_surface. And you are trying to give the returned structure to another malloced pointer.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c [hunk 17]
```diff
@@ -146,140 +146,132 @@ gst_msdkdec_get_oldest_frame (GstVideoDecoder * decoder)
...
-
-    output_state =
-        gst_video_decoder_get_output_state (GST_VIDEO_DECODER (thiz));
-    success =
-        gst_video_frame_map (&i->data, &output_state->info, buffer,
-        GST_MAP_READWRITE);
-    gst_video_codec_state_unref (output_state);
-    if (!success)
-      goto failed_unref_buffer2;
+    return NULL;
```
Comment (tong1wu) [inline subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:264]: I think maybe you should put this in the system memory path.

Comment (mengker33) [reply subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c:264]: As we discussed before, we don't directly use system memory for decoder, so the decoder allocated outbuf is always video memory. We only consider doing gpu2cpu copy when pushing the outbuf forward to downstream, this process can be seen in gst_mskdec_finish_task().

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c (hunk 12 above)
Comment (mengker33) [reply subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c]: Yes, fixed this. Thanks.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c (hunk 13 above)
Comment (mengker33) [reply subprojects/gst-plugins-bad/sys/msdk/gstmsdkallocator_libva.c]: Done.

File: subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c (hunk 14 above)
Comment (mengker33) [reply subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c]: Done, thx.

**Comment by mengker33** [comment]
No more comments. Close.

**Pull request closed without merging**